File: Hard/3068. Find the Maximum Sum of Node Values.cpp
// Greedy (Finding local maxima and minima)
// Time Complexity: O(n)
// Space Complexity: O(1)


class Solution {
public:
   long long maximumValueSum(vector<int>& nums, int k, vector<vector<int>>& edges) {

    long long sum = 0;
    int count = 0;
    int positive_minimum = (1 << 30);
    int negative_maximum = -1 * (1 << 30);

    
    for (int node_value : nums) {
        int operated_node_value = node_value ^ k;
        sum += node_value;
        int net_change = operated_node_value - node_value;

        if (net_change > 0) {
            positive_minimum = min(positive_minimum, net_change);
            sum += net_change;
            count++;
        } else {
            negative_maximum = max(negative_maximum, net_change);
        }
    }


    // if the number of positive net_change values is even return the sum
    if (count % 2 == 0) {
        return sum;
    }

    // otherwise return the maximum of both negative_maximum and positive_minimum
    return max(sum - positive_minimum, sum + negative_maximum);

    }
};

File: Hard/3068. Find the Maximum Sum of Node Values.java
// Greedy (Sorting Based Approch)
// Time complexity: O(n log n)
// Space complexity: O(n)

import java.util.Arrays;

class Solution {
    public long maximumValueSum(int[] nums, int k, int[][] edges) {
        int N = nums.length;
        int[] net_change = new int[N];
        long node_sum = 0;

        for (int i = 0; i < N; i++) {
            net_change[i] = (nums[i] ^ k) - nums[i];
            node_sum += nums[i];
        }

        Arrays.sort(net_change);
        // Reverse the sorted array
        for (int i = 0; i < N / 2; i++) {
            int temp = net_change[i];
            net_change[i] = net_change[N - 1 - i];
            net_change[N - 1 - i] = temp;
        }

        for (int i = 0; i < N; i += 2) {
            // if net_change contains odd number of elements we break the loop
            if (i + 1 == N) {
                break;
            }

            long pair_sum = net_change[i] + net_change[i + 1];
            // include in node_sum if pair_sum is positive (more than 0)
            if (pair_sum > 0) {
                node_sum += pair_sum;
            }
        }

        return node_sum;

    }
}
